Tests/VB/ExpressionTests.cs

```diff
@@ -880,6 +880,61 @@ End Sub
     Public Sub New()
         Dim func As Func(Of String, Boolean) = Function(o) Not String.IsNullOrEmpty(""test"")
     End Sub
+End Class");
+        }
+        [Fact]
+        public async Task EqualsExpression() {
+            await TestConversionCSharpToVisualBasic(
+@"public class TestClass {
+    public TestClass() {
+        int i = 0;
+        int j = 0;
+        string s1 = ""string1"";
+        string s2 = ""string2"";
+        object object1 = s1;
+        object object2 = s2;
+        if(i == j)
+            DoSomething();
+        if(i == s2)
+            DoSomething();
+        if(i == object1)
+            DoSomething();
+        if(s1 == j)
+            DoSomething();
+        if(s1 == s2)
+            DoSomething();
+        if(s1 == object2)
+            DoSomething();
+        if(object1 == j)
+            DoSomething();
+        if(object1 == s2)
+            DoSomething();
+        if(object1 == object2)
+            DoSomething();
+    }
+    public void DoSomething() { }
+}",
+@"Public Class TestClass
+    Public Sub New()
+        Dim i As Integer = 0
+        Dim j As Integer = 0
+        Dim s1 As String = ""string1""
+        Dim s2 As String = ""string2""
+        Dim object1 As Object = s1
+        Dim object2 As Object = s2
+        If i = j Then DoSomething()
+        If i = s2 Then DoSomething()
+        If i = object1 Then DoSomething()
+        If s1 = j Then DoSomething()
+        If s1 = s2 Then DoSomething()
+        If s1 Is object2 Then DoSomething()
+        If object1 = j Then DoSomething()
+        If object1 Is s2 Then DoSomething()
+        If object1 Is object2 Then DoSomething()
+    End Sub
+
+    Public Sub DoSomething()
+    End Sub
```

Comment:
This looks like a good candidate for a self verifying test

Reply:
I'll add one in now

Reply:
I've pushed the tests. Without your change on NodesVisitor.cs:1208 they pass, but with that change they fail.
Therefore I'n not convinced this is the right conversion. Could you take a look and either explain what problem you're fixing, or revert the change if it's wrong? Thanks!

Reply:
**Is** operator compares two objects [by reference](https://docs.microsoft.com/en-gb/dotnet/visual-basic/language-reference/operators/is-operator). It does not perform value comparison.
**=** operator with two strings is compiled to IL using [Microsoft.VisualBasic.Strings.Equals](https://anthonydgreen.net/2019/02/12/exhausting-list-of-differences-between-vb-net-c/#63). So the comparison could be culture-aware. In VB null strings and empty strings are considered equal, and the way to compare a string with null is to use **Is Nothing**. In simple cases it can't be noticed because compiler use String.Intern.

Reply:
Yep so I think we're agreed the runtime equivalent of CSharp "==" in VB is "Is".
If we're not, could you add a test case to the self verifying tests to show where they differ (to pass, the self verifying test must compile and pass in CSharp *and* after conversion to visual basic)

In terms of priorities for conversion, runtime equivalence is priority 1, idiomatic code in the target language is priority 3:
https://github.com/icsharpcode/CodeConverter/blob/master/.github/CONTRIBUTING.md#deciding-what-the-output-should-be
I. E. Code that works the same is key

Note: In CSharp you would use the Equals method to do invariant culture or case insensitive comparison.

Reply:
Nop, in CSharp **==** is compiled to IL to **[System.Runtime]System.String::op_Equality**. It is more equivalent of VB **=**. These simple test passes in CS and not in VB with **Is**:
```
[Fact]
    public void NonInternStringsEqualsOperator() {
        string a = "10";
        string b = 10.ToString();
        Assert.True(a == b);
    }
```
```
Public Sub NonInternStringsEqualsOperator()
        Dim a As String = "10"
        Dim b As String = 10.ToString()
        Assert.[True](a Is b)
End Sub
```

but passes with **=**
```
Public Sub NonInternStringsEqualsOperator()
        Dim a As String = "10"
        Dim b As String = 10.ToString()
        Assert.[True](a = b)
End Sub
```

 Empty string and null string in VB are the same because **Nothing** in VB is not **null** in CSharp, it's default value, for example 0 = Nothing for integers, so in any comparison null and empty string should be the same. The runtime functions also treat null like empty, so it should not be a problem.

Reply:
Ah, my apologies for misunderstanding. I understand now - I had assumed "Is" called op_Equality, but in fact it calls reference equals. This means the implementation in master is flawed as your new interning related test (presumably) highlights.

But my previous tests you removed/changed show the implementation on this PR is also flawed. You'll need to restore those tests to find the correct solution to this problem. I think the solution is to call `Equals(str1, str2)` when both sides are known to be strings.

We should also ensure there's a test case for object == string that preserves the unintended reference equality comparison like this:
https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUBgBjwwGYACAbzzJrI3UutuYHsAjAKwFMBjYMgB5kAvGUwEAdABUWAZWBgAllADmACgCUAbibMaqcWQCeIsgCJxZnbj20DATjVDhoo9t1kAvnk9A